### PR TITLE
Only call +[NSAppearance currentAppearance] on the main thread

### PR DESCRIPTION
**Pull request by amgleitman**
#### Please select one of the following
- [ ] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [ ] I am making a fix / change for the macOS implementation of react-native
- [x] I am making a change required for Microsoft usage of react-native

## Summary

If we try to initialize RCTBridge on a background thread, we end up trying to call `+[NSAppearance currentAppearance]` on that same thread as a result of calling `-[RCTDevLoadingView isDarkModeEnabled]`. However, when we do this, the Main Thread Checker complains that this is a UI function.

To resolve this issue, we ensure that `-[RCTDevLoadingView showOfflineMessage]`, which is making the offending call, does all of its relevant work on the main thread.

### CODE REVIEW

File: React/CoreModules/RCTAppearance.mm
Comment (harrieshin) [inline React/CoreModules/RCTAppearance.mm:106]: this can potentially create a hang. If ui thread kicks off a background thread and waiting on result but the background thread is trying to get the nsappearance on ui thread then it is a deadlock.  what is calling appearance code in the background thread to begin with. are there things can be cached or done async?

Comment (Saadnajmi) [reply React/CoreModules/RCTAppearance.mm:106]: Was about to comment the same. There is a similar helper function in RCTUtils that is explicitly marked as unsafe https://github.com/microsoft/react-native-macos/blob/cf2105fa4915247799f3ce2f030cff5697c7aafc/React/Base/RCTUtils.m#L272

Comment (Saadnajmi) [reply React/CoreModules/RCTAppearance.mm:106]: (For my own knowledge) @harrieshin do you know if iOS has the same constraint about similar API's only being allowed to call on the main thread?

Comment (amgleitman) [reply React/CoreModules/RCTAppearance.mm:106]: If the UI thread is already waiting synchronously for the result of a background thread, that's already a bad thing, isn't it?

Comment (Saadnajmi) [reply React/CoreModules/RCTAppearance.mm:106]: I don't know that there are examples of the UI Thread waiting for a background thread besides at startup when all the native modules are being initialized. I think RN Core does it's best to be asynchronous and have no threads blocking each other, both to avoid hangs and to avoid deadlocks.

Comment (amgleitman) [reply React/CoreModules/RCTAppearance.mm]: In this case, I have an alternative solution that better isolates the problem.

File: React/CoreModules/RCTDevLoadingView.mm
Comment (Saadnajmi) [inline React/CoreModules/RCTDevLoadingView.mm:273]: For my own sake, this is the line that eventually calls `RCTColorSchemePreference(nil)` (Line 289 below), which is checking NSAppearance on a background thread?

Comment (Saadnajmi) [reply React/CoreModules/RCTDevLoadingView.mm:273]: If so, you could maybe also override `requiresMainQueueSetup` to return YES on line 70 ?

Comment (harrieshin) [inline React/CoreModules/RCTDevLoadingView.mm:273]: we should weak reference "self"

Comment (harrieshin) [reply React/CoreModules/RCTDevLoadingView.mm:273]: what caused that change? i would assume checking isdarkmode on ios in the background thread is also wrong. why is it only macOS issue?

File: React/CoreModules/RCTDevLoadingView.mm
Comment (tido64) [inline React/CoreModules/RCTDevLoadingView.mm:269]: Should we be using `RCTExecuteOnMainQueue` here? https://github.com/microsoft/react-native-macos/blob/c647671b0fc781a64f72b1be3378b4359e476bef/React/Base/RCTUtils.h#L35-L37

Comment (amgleitman) [reply React/CoreModules/RCTDevLoadingView.mm]: I didn't realize this was a thing. Nothing else in RCTDevLoadingView uses it, but I can see it being useful here. Might be worth creating a separate issue to track cleaning this up.

File: React/CoreModules/RCTDevLoadingView.mm
Comment (amgleitman) [reply React/CoreModules/RCTDevLoadingView.mm]: Done.

File: React/CoreModules/RCTDevLoadingView.mm
Comment (Saadnajmi) [inline React/CoreModules/RCTDevLoadingView.mm:278]: I would still overall just vote we replace these colors with RCTDynamicColor(light:_, dark:_) and try to push that change upstream. I don't think we should be checking `isDarkModeEnabled` at all, or at least if so, ensure the caller is on the main thread to begin with

Comment (amgleitman) [reply React/CoreModules/RCTDevLoadingView.mm:278]: I feel that for the sake of code cleanliness, we should at least wait until RCTDynamicColor is available on iOS as well before switching patterns (after all, we might as well do this for all instances in this file). I'm not sure why it's only available for macOS now, but I think that the work for doing this is a bit out of scope here.

File: React/CoreModules/RCTDevLoadingView.mm
Comment (amgleitman) [reply React/CoreModules/RCTDevLoadingView.mm]: I'd have to defer to @tido64 on this one.

File: React/CoreModules/RCTDevLoadingView.mm
Comment (Saadnajmi) [reply React/CoreModules/RCTDevLoadingView.mm:278]: It's available on iOS, it's just not a custom class. I would take a look at `RCTConvert.m` and look for the `UIColor` method to see how iOS dynamic colors are made.

Comment (Saadnajmi) [reply React/CoreModules/RCTDevLoadingView.mm:278]: https://github.com/microsoft/react-native-macos/blob/3bcd858a89062d3a77f88e81e1bf6c7624c43794/React/Base/RCTConvert.m#L1051

Comment (Saadnajmi) [reply React/CoreModules/RCTDevLoadingView.mm:278]: Talked offline, this is probably OK to unblock with minimal diffs. I would prefer a better solution pushed upstream but this works.

**Review by Saadnajmi — APPROVED** [review]

File: React/CoreModules/RCTDevLoadingView.mm
Comment (amgleitman) [reply React/CoreModules/RCTDevLoadingView.mm:278]: > Talked offline, this is probably OK to unblock with minimal diffs. I would prefer a better solution pushed upstream but this works.

Yeah, we can probably push a better fix upstream at a later point, but for now I'll resolve this thread.

**Review by harrieshin — APPROVED** [review]

**Pull request merged**
